src/frontend/api.ts

```diff
@@ -143,9 +153,207 @@ export const SERVER_START_TIMEOUT = 30000;
+    put: async (url: string, data: any) => {
+      if (onReady) await onReady();
+      return await req.put(url, { json: data }).json();
+    },
+    post: async (url: string, data: any) => {
+      if (onReady) await onReady();
+      return await req.post(url, { json: data }).json();
+    },
+  };
+}
+
+// TODO: Incorporate server status and making sure it's ready for requests?
+function createMapServerApi() {
+  let mapServerPort: number | undefined;
+  let client: ReturnType<typeof createRequestClient> | undefined;
+
+  // This event occurs whenever the map server's start method is called,
+  // which can happen on app startup but also app resumes
+  nodejs.channel.addListener(
+    "map-server::start",
+    (payload: { port: number }) => {
+      if (mapServerPort !== payload.port) {
+        mapServerPort = payload.port;
+        client = createMapServerClient(mapServerPort);
+      }
+    }
+  );
+
+  function createMapServerClient(port: number) {
+    return createRequestClient({ baseUrl: getBaseUrl(port) });
+  }
+
+  function guaranteeClient() {
+    if (!mapServerPort)
+      throw new Error(
+        "Map server client cannot be used because port is unknown"
+      );
+
+    if (!client) {
+      client = createMapServerClient(mapServerPort);
+    }
+
+    return client;
+  }
+
+  // TODO: Implement addMapServerStateListener and addMapServerErrorListener
+  const mapsApi = {
+    // TODO: Probably should use some kind of status-related implementation similar to how the app server does it
+    ready: () => {
+      // TODO: Rely on the app server's status heartbeat to ping the map server and check its state
+      // This is a temporary measure since there is no server status implemented for the map server right now
+      const appServerStatusSubscription = nodejs.channel.addListener(
+        "status",
+        ({ value }: ServerStatusMessage) => {
+          if (
+            value === STATUS.LISTENING ||
+            value === STATUS.STARTING ||
+            value === STATUS.IDLE
+          ) {
+            nodejs.channel.post("map-server::get-state");
+          }
+        }
+      );
+
+      const readyPromise = new Promise<void>(resolve => {
+        const stateListenerSubscription = nodejs.channel.addListener(
+          "map-server::state",
+          state => {
+            if (state.value === "started") {
+              // @ts-expect-error
+              appServerStatusSubscription.remove();
+              // @ts-expect-error
+              stateListenerSubscription.remove();
+              resolve();
+            }
+          }
+        );
+      });
+
+      const mapServerReadyPromise = promiseTimeout(
+        readyPromise,
+        DEFAULT_TIMEOUT,
+        "Map server start timeout"
+      );
+
+      mapServerReadyPromise.catch(err => {
+        // @ts-expect-error
+        appServerStatusSubscription.remove();
+        bugsnag.notify(err);
+      });
+
+      return mapServerReadyPromise as Promise<void>;
+    },
```

Comment:
Messy, but looks like it will work! I accept full responsibility here for settings this up with this spaghetti tangle between backend and frontend! Will try to get this fixed up soon!